### PR TITLE
CUMULUS-1139 Remove filename property

### PR DESCRIPTION
**Pull request by yjpa7145**
This PR is an incremental step toward simplifying the files schema.  The PR is against the CUMULUS-1139 feature branch, and not against the full master branch.

The `granules` schema contains a `files` field.  The schema for those files previously included a `filename` field, which was an `s3://` URL for the file.  That `filename` field has been removed, and the `bucket` and `filepath` (key) fields should be used instead.  Granule create / update requests that contain a file with a `filename` field will now be rejected, granule files stored in the database will not have a `filename` field, and granule files returned by the API will not have a `filename` field.

### CODE REVIEW

File: example/app/config.yml [hunk 1]
```diff
@@ -300,6 +300,7 @@ kk-uat-deployment:
 mth:
   stackName: mth
   stackNameNoDash: Mth
+  useWorkflowLambdaVersions: false
```
Comment (yjpa7145) [inline example/app/config.yml]: This change is unrelated to the work done in the PR, it is just there to make my deployments more reliable.

File: example/lambdas/snsS3Test/index.js [hunk 2]
```diff
@@ -14,7 +14,8 @@ async function handler(event) {
   const message = JSON.parse(messageString);
   return s3.putObject({
     Bucket: message.cumulus_meta.system_bucket,
-    Key: `${message.meta.stack}/test-output/${message.cumulus_meta.execution_name}.output`
+    Key: `${message.meta.stack}/test-output/${message.cumulus_meta.execution_name}.output`,
+    Body: JSON.stringify(event, null, 2)
```
Comment (yjpa7145) [inline example/lambdas/snsS3Test/index.js]: This change is unrelated to updating the granule files schema, but I found debugging failures of the indexer easier if the content of the event was stored to S3.

File: packages/api/lib/utils.js [hunk 3]
```diff
@@ -71,10 +72,10 @@ function extractDate(payload, dateField) {
  * @returns {Integer} - sum of granule file sizes in bytes
  */
 function getGranuleProductVolume(granuleFiles) {
-  const fileSizes = granuleFiles.map((file) => file.fileSize)
-    .filter((size) => size);
-
-  return fileSizes.reduce((a, b) => a + b);
+  return granuleFiles
```
Comment (yjpa7145) [inline packages/api/lib/utils.js]: The changes made in this file are a refactor and do not contain functionality changes.

File: packages/api/models/files.js [hunk 4]
```diff
@@ -42,19 +22,12 @@ class FileClass extends Manager {
...
-          const extracted = extractFileInfo(file);
-          fileRecords.push({
-            granuleId: granule.granuleId,
-            bucket: extracted.bucket,
-            key: extracted.key
-          });
-        }
-      });
-    }
+    const fileRecords = (granule.files || [])
```
Comment (yjpa7145) [inline packages/api/models/files.js]: Refactoring.

File: packages/api/models/files.js [hunk 5]
```diff
@@ -68,18 +41,11 @@ class FileClass extends Manager {
...
-        if (file.filename) {
-          const extracted = extractFileInfo(file);
-          fileRecords.push({
-            bucket: extracted.bucket,
-            key: extracted.key
-          });
-        }
-      });
-    }
+    const fileRecords = (granule.files || [])
```
Comment (yjpa7145) [inline packages/api/models/files.js]: Refactoring.

File: packages/api/models/files.js [hunk 6]
```diff
@@ -96,31 +62,19 @@ class FileClass extends Manager {
...
    */
   async deleteFilesAfterCompare(newGranule, oldGranule) {
-    if (oldGranule && oldGranule.files) {
-      const currentFiles = {};
-      if (newGranule.files) {
-        newGranule.files.forEach((file) => {
-          currentFiles[file.filename] = file;
-        });
-      }
+    const buildFileId = (f) => `${f.bucket}/${f.filepath}`;
```
Comment (yjpa7145) [inline packages/api/models/files.js]: Refactoring.

File: packages/common/aws.js [hunk 7]
```diff
@@ -299,8 +309,21 @@ exports.s3PutObjectTagging = (bucket, key, tagging) =>
 * @param {string} key - key for object (filepath + filename)
 * @returns {Promise} - returns response from `S3.getObject` as a promise
 **/
-exports.getS3Object = (bucket, key) =>
-  exports.s3().getObject({ Bucket: bucket, Key: key }).promise();
+exports.getS3Object = async (bucket, key) => {
```
Comment (yjpa7145) [inline packages/common/aws.js]: If there is an error in an S3 API call, the stack trace gives no information about what the failing operation was or where it occurred.  This code provides a more useful stack trace when someone calls `getS3Object()`, which actually shows what the call was, where it was, and what the params were.  Makes debugging a _lot_ easier.

File: example/spec/helpers/testUtils.js [hunk 8]
```diff
@@ -197,7 +197,17 @@ async function redeploy(config, options = {}) {
...
 
 async function getFileMetadata(file) {
-  const { Bucket, Key } = parseS3Uri(file.filename);
+  let Bucket;
+  let Key;
+  if (file.bucket && file.filepath) {
+    Bucket = file.bucket;
+    Key = file.filepath;
+  }
+  else {
```
Comment (jennyhliu) [inline example/spec/helpers/testUtils.js]: https://github.com/nasa/cumulus/pull/800/files#diff-59617f11d03d238d58e5a4255941ea08R215
Why does this method still return `filename: file.filename` if `file.filename` may not be set.
Should we remove `else` block?

File: packages/ingest/granule.js [hunk 9]
```diff
@@ -714,22 +714,27 @@ async function moveGranuleFile(source, target, options) {
...
+      source = {
+        Bucket: file.bucket,
+        Key: file.filepath
       };
-
-      return { source, target, file };
     }
-    return { source: null, target: null, file };
+    else {
+      source = aws.parseS3Uri(file.filename);
```
Comment (jennyhliu) [inline packages/ingest/granule.js]: `file.filename` is still there?

File: tasks/post-to-cmr/index.js [hunk 10]
```diff
@@ -78,7 +79,15 @@ async function postToCMR(event) {
   granules.forEach((granule) => {
     allGranules[granule.granuleId] = granule;
     granule.files.forEach((file) => {
-      allFiles.push(file.filename);
+      if (file.bucket && file.filepath) {
+        allFiles.push(buildS3Uri(file.bucket, file.filepath));
+      }
+      else if (file.filename) {
```
Comment (jennyhliu) [inline tasks/post-to-cmr/index.js]: It's confusing.  If `granule.files` no longer has `filename`, why do we still check `file.bucket && file.filepath` and `file.filename`?  Should `file.bucket && file.filepath` always present?

File: packages/api/lib/granuleUtils.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+'use strict';
+
+const { parseS3Uri } = require('@cumulus/common/aws');
+
+const getKeyFromFile = (file) => {
+  if (file.filename) return parseS3Uri(file.filename).Key;
```
Comment (markdboyd) [inline packages/api/lib/granuleUtils.js]: If this is being removed, why are we still checking for it?

File: packages/ingest/granule.js [hunk 12]
```diff
@@ -766,12 +770,26 @@ async function moveGranuleFiles(sourceFiles, destinations) {
...
+      fileKey = parsed.Key;
+    }
+
     processedFiles.push({
-      bucket: parsed.Bucket,
-      filepath: parsed.Key,
-      filename: file.filename,
+      bucket: fileBucket,
+      filepath: fileKey,
+      filename: aws.buildS3Uri(fileBucket, fileKey),
```
Comment (markdboyd) [inline packages/ingest/granule.js]: Why are we continuing to build this property if we're deprecating it?

File: packages/cmrjs/cmr-utils.js [hunk 13]
```diff
@@ -379,7 +380,9 @@ function mergeURLs(original, updated, removed = []) {
 async function updateUMMGMetadata(cmrFile, files, distEndpoint, buckets) {
   const newURLs = constructOnlineAccessUrls(files, distEndpoint, buckets);
   const removedURLs = onlineAccessURLsToRemove(files, buckets);
-  const metadataObject = await metadataObjectFromCMRJSONFile(cmrFile.filename);
+  const filename = cmrFile.filename
```
Comment (markdboyd) [inline packages/cmrjs/cmr-utils.js]: Still a bit confused why we're checking the `filename` property?

File: packages/api/lib/granuleUtils.js (hunk 11 above)
Comment (markdboyd) [reply packages/api/lib/granuleUtils.js]: I think I see. This is a temporary measure as part of the converting files from a cumulus message to the new API format.

What will still have the `filename` property that makes this code necessary?

File: packages/ingest/granule.js [hunk 14]
```diff
@@ -766,12 +770,26 @@ async function moveGranuleFiles(sourceFiles, destinations) {
         });
       });
     }
+
+    let fileBucket;
+    let fileKey;
+    if (file.bucket && file.filepath) {
```
Comment (ifestus) [inline packages/ingest/granule.js]: This same `if, else` block is used a couple times throughout your changes. I'm wondering if this might be a good block to move into a separate function. Then you'd have something like `const { fileBucket, fileKey } = myCoolFunctionTM(file);`.

File: packages/api/lib/granuleUtils.js (hunk 11 above)
Comment (ifestus) [reply packages/api/lib/granuleUtils.js]: I don't know if this is the reason, but we have some tests that expect this value to still be alive. There might also be some fixtures/ancillary setup data that use it.

File: packages/api/lib/testUtils.js [hunk 15]
```diff
@@ -42,23 +43,27 @@ async function deleteAliases() {
   })));
 }
 
+function fakeFileFactory(params = {}) {
+  const name = randomId('name');
+
+  return {
+    bucket: randomString(),
+    name,
```
Comment (ifestus) [inline packages/api/lib/testUtils.js]: Maybe a little bit out of scope, but are `name` and `filepath` always expected to have the same value? I'm wondering why we keep both around.

File: packages/api/models/files.js (hunk 4 above)
Comment (ifestus) [reply packages/api/models/files.js]: Nice :+1:

**Comment by yjpa7145** [comment]
To bring some clarity to the whole, "why is filename still around in these places" discussion:

This PR is updating the API so that it will no longer store the `filename` field, and will no longer return it.  There are some other places in our code (mostly in tasks) that still use the `filename` field in their input, and produce it in their output.  Updating all of those references was out of the scope of this ticket.

There is some code in the `@cumulus/ingest` package and in the `@cumulus/cmrjs` package that take a file as an argument.  Those functions that are taking a file as an argument are called from tasks as well as from the API.  Since tasks have not been updated, that code needs to be able to handle an incoming file argument that contains a `filename` field (and possibly not a `bucket` or `filepath` field).  That code also needs to be able to handle a file argument coming from the API, which will contain a `bucket` and `filepath`, but should not contain a `filename`.

Since people have brought it up, I'm taking a closer look at the places that we're still using `filename` to make sure that it is actually still required.

File: packages/api/lib/granuleUtils.js (hunk 11 above)
Comment (yjpa7145) [reply packages/api/lib/granuleUtils.js]: This comment talks about why there are still instances of `filename`: https://github.com/nasa/cumulus/pull/800#issuecomment-462491409

That being said, I've taken a closer look at where this instance of `filename` is used, and I _think_ it can be removed.  Waiting on one last test build ...

File: packages/api/lib/testUtils.js (hunk 15 above)
Comment (yjpa7145) [reply packages/api/lib/testUtils.js]: They will not always be the same.  From the CNM docs:

> Some providers will only create hashes (to maximize read/write performance to/from S3 buckets). as object keys. We need a way to define what the object name should be, and explicitly told (not guessing).

File: packages/cmrjs/cmr-utils.js (hunk 13 above)
Comment (yjpa7145) [reply packages/cmrjs/cmr-utils.js]: This is one of the instances where, depending on where it's being called from, you're going to have different types of files.

When a granule is moved using the API, the files that make it to here will have `bucket` and `filepath` fields, but no `filename`.  This code is also called as part of the `move-granules` task, which will send it a `filename`.

File: packages/api/lib/granuleUtils.js (hunk 11 above)
Comment (yjpa7145) [reply packages/api/lib/granuleUtils.js]: And ... yeah, we still need `filename` here.  The `filename` field seems to be the only reliable way to determine a file's key from the output of the `SyncGranule` task, which is needed by `packages/api/es/indexer.js`.

File: example/spec/helpers/testUtils.js (hunk 8 above)
Comment (yjpa7145) [reply example/spec/helpers/testUtils.js]: The `getFileMetadata` function is used for files returned as task output and for files returned from the API.  Files from the API will have a `bucket` and `filepath`, while a file returned as the output of a task will have a `filename`.

* Verifying task output here - https://github.com/nasa/cumulus/blob/7b4793988a98d68e7fbad9b0d699798ae4c3686c/example/spec/parallel/ingestGranule/IngestGranuleDuplicateHandlingSpec.js#L113
* Verifying response from the API here - https://github.com/nasa/cumulus/blob/5e868fa749ffe5b1c8d67ea0261f0358d1d3825c/example/spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js#L464

File: packages/ingest/granule.js (hunk 9 above)
Comment (yjpa7145) [reply packages/ingest/granule.js]: Fixing this instance.

**Comment by markdboyd** [comment]
I understand not including the refactor of `tasks`, though I don't love that it puts us in a place where we're adding temporary glue code (e.g. `cumulusMessageFileToAPIFile`) that I think we'll want to delete once the refactor of the tasks is complete? Perhaps I'm misunderstanding something though.

Either way, I don't have any problems with the code changes themselves, as long as the tests pass

**Comment by yjpa7145** [comment]
> I understand not including the refactor of the tasks, though I don't love that it puts us in a place where we're adding temporary glue code (e.g. `cumulusMessageFileToAPIFile`) that I think we'll want to delete once the refactor of the tasks is complete? Perhaps I'm misunderstanding something though.

Typically I would agree with you but, having been stuck hip deep in this schema change for months, I've definitely learned that you need to tackle the smallest piece that you can.  We can certainly look at filing tickets to change the task outputs later.  If we do that, we're also going to need to update the documentation that describes the format of the message to write to the `sftracker` SNS topic.

File: packages/ingest/granule.js (hunk 14 above)
Comment (yjpa7145) [reply packages/ingest/granule.js]: We're doing the same check in a few different places, but taking different actions.  In some cases we're returning bucket / key, and others we're returning the S3 URL.  I'm hoping that, soon, we'll be able to just take these conditionals entirely out.  That'll have to wait until after we change the format of messages that are sent to the `sftracker` SNS topic, though.

File: tasks/post-to-cmr/index.js (hunk 10 above)
Comment (yjpa7145) [reply tasks/post-to-cmr/index.js]: It depends on where the input to the `post-to-cmr` task is coming from.  If it's coming from `discover-granules` (or any other task), it's going to have a `filename` field.  If the granules being used as input came from the API then they will have `bucket` and `filepath` fields, but no `filename`.

File: example/app/config.yml (hunk 1 above)
Comment (Jkovarik) [reply example/app/config.yml]: @yjpa7145  Should we be concerned that you needed to do this?

Comment (flamingbear) [reply example/app/config.yml]: I do it when I have to make bunches of changes to a single lambda and want to deploy just that one.
